### PR TITLE
Added Reader and RWS instances for MonadRec

### PR DESCRIPTION
**Pull request by badly-drawn-wizards**
Easy additions. However the RWS recursion is unacceptably slow, and I am unsure if that is my fault.

### CODE REVIEW

File: src/Control/Monad/Rec/Class.purs [hunk 1]
```diff
@@ -116,10 +121,17 @@ instance monadRecWriterT :: (Monoid w, MonadRec m) => MonadRec (WriterT w m) whe
       Left a -> Left (Tuple a (w <> w1))
       Right b -> Right (Tuple b (w <> w1))) (Tuple a mempty)
 
+instance monadRecRWS :: (Monoid w, MonadRec m) => MonadRec (RWST r w s m) where
+  tailRecM f a = RWST $ \r s -> tailRecM (\{state:s,result:a,log:w} -> do
+    {state:s1, log:w1, result: m} <- runRWST (f a) r s
+    return case m of
+      Left a -> Left ({state:s1,result:a,log:w<>w1})
```
Comment (paf31) [inline src/Control/Monad/Rec/Class.purs]: Could you please break up the records with spaces here?

**Comment by paf31** [comment]
Looks good. As for the speed, how slow are we talking? I suspect it is something to do with creating lots of records. Multi-argument functions would probably be a better solution, but there is no `tailRec` for the `FnN` types. If we want to address the speed issues (and possibly speed up `StateT` and others too), we should probably add multi-arg versions of `tailRec` and delegate to those instead.

**Comment by badly-drawn-wizards** [comment]
I'm talking about on the order of a couple of seconds as opposed to the split second times I get for the other tests running the same number of iterations.

**Comment by paf31** [comment]
Hmm. Maybe it's worth trying the multi-argument function way then. _Edit_: obviously the return value will still need to be a record, but maybe that helps a bit. Also, you could try using a data constructor. Maybe `RWS` is just generally slow for this reason, or maybe it's something else.

**Comment by badly-drawn-wizards** [comment]
I think if RWS is to improve sufficiently in performance it must be done by making it use a data constructor in purescript-transformers. Not to say that it can't be improved here too, but that is the smaller issue. 
I have tried using nested tuples and it's still bad. I haven't tried data constructors yet, but I suspect it will be the same.

**Comment by paf31** [comment]
These instances would be in `transformers` now anyway, so I'll close this. Thanks for the work on this though.

**Pull request closed without merging**